test/integration/eds_integration_test.cc

```diff
@@ -437,5 +438,486 @@ TEST_P(EdsIntegrationTest, StatsReadyFilter) {
+    setUpstreamCount(4);
+    setUpstreamProtocol(Http::CodecType::HTTP2);
+
+    config_helper_.addConfigModifier(
+        [this, http_active_hc, num_clusters_to_add,
+         ignore_new_hosts_until_first_hc](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+          // Add a static EDS cluster.
+          auto* eds_cluster = bootstrap.mutable_static_resources()->add_clusters();
+          eds_cluster->MergeFrom(bootstrap.static_resources().clusters()[0]);
+          eds_cluster->set_name("eds_cluster");
+          eds_cluster->mutable_load_assignment()->set_cluster_name("eds_cluster");
+          ConfigHelper::setHttp2(*eds_cluster);
+          // Remove the static cluster (cluster_0) and set up CDS.
+          bootstrap.mutable_dynamic_resources()->mutable_cds_config()->set_resource_api_version(
+              envoy::config::core::v3::ApiVersion::V3);
+          bootstrap.mutable_dynamic_resources()
+              ->mutable_cds_config()
+              ->mutable_path_config_source()
+              ->set_path(cds_helper_.cds_path());
+          bootstrap.mutable_static_resources()->mutable_clusters()->erase(
+              bootstrap.mutable_static_resources()->mutable_clusters()->begin());
+          for (uint32_t i = 0; i < num_clusters_to_add; ++i) {
+            auto cluster_name = "cluster_" + std::to_string(i);
+            auto cluster_to_add =
+                buildCluster(cluster_name, http_active_hc, ignore_new_hosts_until_first_hc);
+            envoy::config::endpoint::v3::ClusterLoadAssignment cla_to_add;
+            cla_to_add.set_cluster_name(cluster_name);
+            clas_to_add_.push_back(cla_to_add);
+            clusters_to_add_.push_back(cluster_to_add);
+            resource_names_.push_back(cluster_name);
+          }
+          cds_helper_.setCds(clusters_to_add_);
+        });
+    // Set validate_clusters to false to allow us to reference a CDS cluster.
+    config_helper_.addConfigModifier(
+        [](envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager&
+               hcm) { hcm.mutable_route_config()->mutable_validate_clusters()->set_value(false); });
+    defer_listener_finalization_ = true;
+    HttpIntegrationTest::initialize();
+    // Add the assignment and localities.
+    cluster_load_assignment_.set_cluster_name("cluster_0");
+    acceptXdsConnection();
+    for (uint32_t i = 0; i < num_clusters_to_add; ++i) {
+      EXPECT_TRUE(compareDiscoveryRequest(
+          Config::TypeUrl::get().ClusterLoadAssignment, "",
+          {resource_names_.begin(), resource_names_.begin() + i + 1}, {}, {}, true));
+    }
+    sendDiscoveryResponse<envoy::config::endpoint::v3::ClusterLoadAssignment>(
+        Config::TypeUrl::get().ClusterLoadAssignment, clas_to_add_, clas_to_add_, {},
+        std::to_string(eds_version_));
+    // Receive EDS ack.
+    EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().ClusterLoadAssignment,
```

Comment:
I may be missing something here, but shouldn't this test file verify that both the multiplexed and non-multiplexed are working correctly?
If so, shouldn't there be multiple xds streams, and sending updates on multiple streams (in the case of non-multiplexed), and use of a single stream in the multiplexed case?
Please let me know if I'm missing something.

Reply:
That is correct. I intentionally omitted tests for multiple streams/clusters with no multiplexing, because the code is async and is not possible to predict the order of discovery requests. Is there is an established way to test out of order expected set of discovery requests/responses?

Reply:
That's a great question.
I don't think there is something ready. I've implemented this once for delta-xDS in a previous PR (https://github.com/envoyproxy/envoy/pull/18195/commits/20b1674193e62b6087dd13340b7bc91c03de6301#diff-5227f6492ddd6badd76af87ce9f5f1c8acc349eed191e65a2bc05ab5d43b65f6R643) but eventually it wasn't needed for the tests.